components/data-feed-service/server/server.go

```diff
@@ -0,0 +1,182 @@
+	log.Info("NewDatafeedServer")
+	secretsConn, err := connFactory.Dial("secrets-service", config.SecretsConfig.Target)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not connect to secrets-service")
+	}
+	return &DatafeedServer{
+		db:      db,
+		health:  health.NewService(),
+		secrets: secrets.NewSecretsServiceClient(secretsConn),
+	}, nil
+}
+
+// Add a new destination
+func (datafeedServer *DatafeedServer) AddDestination(ctx context.Context, destination *datafeed.Destination) (*datafeed.DestinationResponse, error) {
+	log.Infof("AddDestination %s", destination)
+	response := &datafeed.DestinationResponse{Success: false}
+	success, err := datafeedServer.db.AddDestination(destination)
+	response.Success = success
+	if err != nil {
+		return response, errorutils.FormatErrorMsg(err, "")
+	}
+
+	return response, nil
+}
+
+func (datafeedServer *DatafeedServer) TestDestination(ctx context.Context, request *datafeed.URLValidationRequest) (*datafeed.DestinationResponse, error) {
+	log.Infof("TestDestination %s", request)
+	response := &datafeed.DestinationResponse{Success: false}
+	// http client to endpoint {text: "TEST: Successful validation completed by Automate"}
+	// if it's secret - get the credentials
+	// otherwise use passwd
+	username := ""
+	password := ""
+	var err error
+	url := request.Url
+	switch request.Credentials.(type) {
+	case *datafeed.URLValidationRequest_UsernamePassword:
+		username = request.GetUsernamePassword().GetUsername()
+		password = request.GetUsernamePassword().GetPassword()
+	case *datafeed.URLValidationRequest_SecretId:
+		secretId := request.GetSecretId().GetId()
+		// call secrets api
+		username, password, err = service.GetCredentials(context.Background(), datafeedServer.secrets, secretId)
+		if err != nil {
+			return response, err
+		}
+	}
+
+	messageBytes, err := json.Marshal(map[string]string{
+		"text": "TEST: Successful validation completed by Automate",
+	})
+	if err != nil {
+		log.Errorf("Error creating json bytes %v", err)
+		return response, err
+	}
+
+	var contentBuffer bytes.Buffer
+	zip := gzip.NewWriter(&contentBuffer)
+	_, err = zip.Write(messageBytes)
+	if err != nil {
+		return response, err
+	}
+	err = zip.Close()
+	if err != nil {
+		return response, err
+	}
+
+	httpRequest, err := http.NewRequest("POST", url, &contentBuffer)
+	if err != nil {
+		log.Error("Error creating request")
+		return response, err
+	}
+	if username != "" && password != "" {
+		httpRequest.SetBasicAuth(username, password)
+	}
+	httpRequest.Header.Add("Content-Type", "application/json")
+	httpRequest.Header.Add("Content-Encoding", "gzip")
+	httpRequest.Header.Add("Accept", "application/json")
+	client := http.Client{}
+	httpResponse, err := client.Do(httpRequest)
+	if err != nil {
+		log.Errorf("Error sending test message %v", err)
+		return response, err
+	} else {
+		log.Infof("Test data posted to %v, Status %v", url, httpResponse.Status)
+	}
+	if httpResponse.StatusCode != http.StatusCreated {
```

Comment:
Why is only a 201 the correct response? When testing a new data feed from the UI I am seeing "We were unable to connect to your destination" because the test server is returning a 200.

Reply:
The service now code returns 201 as the import is asynchronous and we have only created the import records, these need processed before the CMDB tables are populated. Since this isn't restricted to service now that's a very valid point. I'll change to a 200 and return 200 from service now in the event of a test message.